libusb/descriptor.c

```diff
@@ -1001,6 +1006,62 @@ int API_EXPORTED libusb_get_ss_usb_device_capability_descriptor(
 	return LIBUSB_SUCCESS;
 }
 
+int API_EXPORTED libusb_get_ssplus_usb_device_capability_descriptor(
+	libusb_context *ctx,
+	struct libusb_bos_dev_capability_descriptor *dev_cap,
+	struct libusb_ssplus_usb_device_capability_descriptor **ssplus_usb_device_cap) 
+{
+	struct libusb_ssplus_usb_device_capability_descriptor *_ssplus_usb_device_cap;
+	uint16_t i;
+	
+	if (dev_cap->bDevCapabilityType != LIBUSB_BT_SUPERSPEED_PLUS_CAPABILITY) {
+		usbi_err(ctx, "unexpected bDevCapabilityType 0x%x (expected 0x%x)",
+			 dev_cap->bDevCapabilityType,
+				 LIBUSB_BT_SUPERSPEED_PLUS_CAPABILITY);
+		return LIBUSB_ERROR_INVALID_PARAM;
+	} else if (dev_cap->bLength < LIBUSB_BT_SSPLUS_USB_DEVICE_CAPABILITY_SIZE) {
+		usbi_err(ctx, "short dev-cap descriptor read %u/%d",
+			 dev_cap->bLength, LIBUSB_BT_SSPLUS_USB_DEVICE_CAPABILITY_SIZE);
+		return LIBUSB_ERROR_IO;
+	}
+
+	_ssplus_usb_device_cap = malloc(sizeof(*_ssplus_usb_device_cap));
+	if (!_ssplus_usb_device_cap)
+		return LIBUSB_ERROR_NO_MEM;
+
+	// We can only parse the non-variable size part of the SuperSpeedPlus descriptor.
+	parse_descriptor(dev_cap, "bbbbiww", _ssplus_usb_device_cap);
+	
+	// Now, parse the SublinkSpeed attributes which size varies depending on bmAttributes. Let's start by retrieving
+	// how many attributes are there.
+	uint16_t num_sublink_attributes = (_ssplus_usb_device_cap->bmAttributes & LIBUSB_SSPLUS_ATTRIBUTE_SSCA_MASK) + 1;
+	
+	// Check that we have enough to read all the sublink attributes
+	if (dev_cap->bLength < LIBUSB_BT_SSPLUS_USB_DEVICE_CAPABILITY_SIZE + num_sublink_attributes * sizeof(uint32_t)) {
+		usbi_err(ctx, "short ssplus capability descriptor, unable to read sublinks: Not enough data");
+		return LIBUSB_ERROR_IO;
+	}
+
+	// Finally, read the attributes
+	_ssplus_usb_device_cap->bmSublinkSpeedAttr = malloc(num_sublink_attributes * sizeof(uint32_t));
+	uint32_t* attributes_pointer = (uint32_t*)(((uint8_t*)dev_cap) + LIBUSB_BT_SSPLUS_USB_DEVICE_CAPABILITY_SIZE);
+	for(i=0 ; i < num_sublink_attributes ; i++) {
+		_ssplus_usb_device_cap->bmSublinkSpeedAttr[i] = *attributes_pointer++;
+	}
+
+	*ssplus_usb_device_cap = _ssplus_usb_device_cap;
+	return LIBUSB_SUCCESS;
+}
+
+void API_EXPORTED libusb_free_ssplus_usb_device_capability_descriptor(
+	struct libusb_ssplus_usb_device_capability_descriptor *ssplus_usb_device_cap)
+{
+	free(ssplus_usb_device_cap->bmSublinkSpeedAttr);
```

Comment:
I would add a null check on ssplus_usb_device_cap here. It is ok to call free(NULL) but this dereferences a potentially null pointer.

Reply:
Indeed other functions named `libusb_free...` allow for NULL being passed.